plugins/meta/portmap/portmap.go

```diff
@@ -0,0 +1,266 @@
+		rule := []string{
+			"-p", entry.Protocol,
+			"--dport", strconv.Itoa(entry.HostPort)}
+
+		if entry.HostIP != "" {
+			rule = append(rule,
+				"-d", entry.HostIP)
+		}
+
+		rule = append(rule,
+			"-j", "DNAT",
+			"--to-destination", fmtIpPort(containerIP, entry.ContainerPort))
+
+		out = append(out, rule)
+	}
+	return out
+}
+
+// genToplevelSnatChain creates the top-level summary snat chain.
+// IMPORTANT: do not change this, or else upgrading plugins will require
+// manual intervention
+func genToplevelSnatChain(isV6 bool) chain {
+	return chain{
+		table: "nat",
+		name:  TopLevelSNATChainName,
+		entryRule: []string{
+			"-s", localhostIP(isV6),
+			"!", "-d", localhostIP(isV6),
+		},
+		entryChains: []string{"POSTROUTING"},
+	}
+}
+
+// genSnatChain creates the snat (localhost) chain for this container.
+func genSnatChain(netName, containerID string) chain {
+	name := formatChainName("SN-", netName, containerID)
+	comment := fmt.Sprintf(`snat name: "%s" id: "%s"`, netName, containerID)
+
+	return chain{
+		table: "nat",
+		name:  name,
+		entryRule: []string{
+			"-m", "comment",
+			"--comment", comment,
+		},
+		entryChains: []string{TopLevelSNATChainName},
+	}
+}
+
+// snatRules sets up masquerading for connections to localhost:hostport,
+// rewriting the source so that returning packets are correct.
+func snatRules(entries []PortMapEntry, containerIP net.IP) [][]string {
+	isV6 := (containerIP.To4() == nil)
+
+	out := make([][]string, 0, len(entries))
+	for _, entry := range entries {
+		out = append(out, []string{
+			"-p", entry.Protocol,
+			"-s", localhostIP(isV6),
+			"-d", containerIP.String(),
+			"--dport", strconv.Itoa(entry.ContainerPort),
+			"-j", "MASQUERADE",
+		})
+	}
+	return out
+}
+
+// enableLocalnetRouting tells the kernel not to treat 127/8 as a martian,
+// so that connections with a source ip of 127/8 can cross a routing boundary.
+func enableLocalnetRouting(ifName string) error {
+	routeLocalnetPath := "net.ipv4.conf." + ifName + ".route_localnet"
+	_, err := sysctl.Sysctl(routeLocalnetPath, "1")
+	return err
+}
+
+// unforwardPorts deletes any iptables rules created by this plugin.
+// It should be idempotent - it will not error if the chain does not exist.
+func unforwardPorts(config *PortMapConf) error {
+	// Teardown v4
+	ipt, err := iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	if err != nil {
+		return err
+	}
+	dnatChain := genDnatChain(config.Name, config.ContainerID, nil)
+	if err := dnatChain.teardown(ipt); err != nil {
+		return fmt.Errorf("could not teardown ipv4 dnat: %v", err)
+	}
+	snatChain := genSnatChain(config.Name, config.ContainerID)
+	if err := snatChain.teardown(ipt); err != nil {
+		return fmt.Errorf("could not teardown ipv4 snat: %v", err)
+	}
+
+	ip6t, err := iptables.NewWithProtocol(iptables.ProtocolIPv6)
+	if err != nil {
+		return err
+	}
+
+	if err := dnatChain.teardown(ip6t); err != nil {
+		return fmt.Errorf("could not teardown ipv6 dnat: %v", err)
```

Comment:
I think that's safe to assume, but it's also possible that that the ip6tables tool is present but the required kernel support is not, and we don't want to fail in that case either I think.

I'm not sure there's a good solution - ideally we'd be able to know if ip6tables rules were actually created for this container.

I'd be happy with this being "best-effort" - if it can't init the table then it logs a warning and carries on.

I'd also be happy with a global "ipv6_enabled" option,  but that means people deploying this will need to make a decision.  Seems like ipv6_enabled could even default to false, and the plugin could return an error if it gets ipv6 mappings.